### PR TITLE
Axis decorators for rotated and offset labels

### PR DESCRIPTION
**Pull request by andy-lee-eng**
New decorators for axis components.

**axisLabelRotate** automatically rotates labels when there isn't enough room to display them side-by-side, or can be configured to use a pre-set rotation.

**axisLabelOffset** automatically splits labels into multiple rows on a horizontal axis (columns on vertical), or can be configured to use a pre-set depth (number of rows/columns).

Examples and documentation included.

### CODE REVIEW

File: packages/d3fc-axis/README.md [hunk 1]
```diff
@@ -128,3 +122,40 @@ const axis = fc.axisBottom(scale)
...
+![](screenshots/ordinal.png)
+
+<a name="axisOrdinalBottom" href="#axisOrdinalBottom">#</a> fc.**axisOrdinalBottom**()  
+<a name="axisOrdinalTop" href="#axisOrdinalTop">#</a> fc.**axisOrdinalTop**()  
+<a name="axisOrdinalLeft" href="#axisOrdinalLeft">#</a> fc.**axisOrdinalLeft**()  
+<a name="axisOrdinalRight" href="#axisOrdinalRight">#</a> fc.**axisOrdinalRight**()  
+
+A drop-in replacement for D3 axis, see the [D3 documentation for API reference](https://github.com/d3/d3-axis#api-reference).
+
+### Axis Label Offset
```
Comment (ColinEberhardt) [inline packages/d3fc-axis/README.md:148]: @DevAndyLee can you please add similar docs for the `axisLabelRotate` adapter?

File: packages/d3fc-axis/README.md [hunk 2]
```diff
@@ -128,3 +122,40 @@ const axis = fc.axisBottom(scale)
...
+
+![](screenshots/offset.png)
+
+<a name="axisLabelOffset" href="#axisLabelOffset">#</a> fc.**axisLabelOffset**()
```
Comment (ColinEberhardt) [inline packages/d3fc-axis/README.md]: Ooops - just spotted that this is not correct. The function takes an arg, the axis to adapt

Comment (ColinEberhardt) [reply packages/d3fc-axis/README.md]: assign #self

File: packages/d3fc-axis/src/axisLabelRotate.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    let labelRotate = 'auto';
+    let decorate = () => { };
+
+    const measureLabels = s => {
+        const labels = adaptee.scale().domain();
+
+        // Use a test element to measure the text in the axis SVG container
+        const tester = s.append('text');
+        const labelHeight = tester.text('Ty').node().getBBox().height;
```
Comment (ColinEberhardt) [inline packages/d3fc-axis/src/axisLabelRotate.js]: In the label offset adapter I [measure the height](https://github.com/d3fc/d3fc/pull/1263/files#diff-2c22a2488cd6e13efc5f0d9087a94ec7R14) - would that work here too?

File: packages/d3fc-axis/src/axisOrdinal.js [hunk 4]
```diff
@@ -40,7 +40,9 @@ const axisOrdinal = (orient, scale) => {
     const base = axisBase(orient, scale, {labelOffset, tickPath});
 
     const axis = (selection) => {
-        return base(selection);
+        // if (!rotation) calculateRotation(selection);
```
Comment (ColinEberhardt) [inline packages/d3fc-axis/src/axisOrdinal.js]: oops - my bad. I'll fix this.

File: packages/d3fc-axis/src/axisLabelRotate.js (hunk 3 above)
Comment (andy-lee-eng) [reply packages/d3fc-axis/src/axisLabelRotate.js]: I've made the two measuring functions the same now, and extracted the shared code.

File: packages/d3fc-axis/README.md (hunk 1 above)
Comment (andy-lee-eng) [reply packages/d3fc-axis/README.md:148]: Done

**Pull request merged**